### PR TITLE
Adding models for Zuul tests.

### PR DESCRIPTION
**Pull request by jsanemet**
On this pull request:
- Added 'Test' model for generic test cases.
- Added 'AnsibleTest' model for tests based on Ansible tasks.
- Added 'TempestTest' model for test best on Tempest test cases.

The models are not integrated yet, just defined.

### CODE REVIEW

**Review by bregman-arie — COMMENTED** [review]
For me, a test should be the same whether it's running on Jenkins or Zuul. In both systems, a test has the same properties. So why are we creating Zuul-specific test models?

File: tests/unit/models/ci/zuul/tests/test_ansible.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    def test_equality_by_type(self):
+        """Checks that two models are no the same if they are of different
+        type.
+        """
+        model = AnsibleTest()
+        other = Mock()
+
+        self.assertNotEqual(other, model)
+
+    def test_equality_by_reference(self):
```
Comment (amolkahat) [inline tests/unit/models/ci/zuul/tests/test_ansible.py:70]: ditto

File: tests/unit/models/ci/zuul/tests/test_ansible.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        """Checks that a model is equal to itself.
+        """
+        model = AnsibleTest()
+
+        self.assertEqual(model, model)
+
+    def test_equality_by_contents(self):
```
Comment (amolkahat) [inline tests/unit/models/ci/zuul/tests/test_ansible.py:77]: ditto

File: tests/unit/models/ci/zuul/tests/test_tempest.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    def test_equality_by_type(self):
+        """Checks that two models are no the same if they are of different
+        type.
+        """
+        model = TempestTest()
+        other = Mock()
+
+        self.assertNotEqual(other, model)
+
+    def test_equality_by_reference(self):
```
Comment (amolkahat) [inline tests/unit/models/ci/zuul/tests/test_tempest.py:65]: ditto

Comment (amolkahat) [inline tests/unit/models/ci/zuul/tests/test_tempest.py:65]: Same here as well.

**Review by amolkahat — CHANGES_REQUESTED** [review]

File: tests/unit/models/ci/zuul/test_test.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        """
+        model = Test(TestKind.UNKNOWN, Test.Data())
+        other = Mock()
+
+        self.assertNotEqual(other, model)
+
+    def test_equality_by_reference(self):
+        """Checks that a model is equal to itself.
+        """
+        model = Test(TestKind.UNKNOWN, Test.Data())
```
[the remark this answers is not in the document]
Comment (jsanemet) [reply tests/unit/models/ci/zuul/test_test.py:61]: The classes that I have added overwrite the '__eq__' operator, so these tests make sure that I am keeping the condition you mention true. That is why I have a test explicitly for so.

**Comment by jsanemet** [comment]
@bregman-arie The test models are very similar in both systems, but I have given Zuul its own model because I have had a bad experience reusing code from Jenkins before. Zuul always shows some condition that diverges from Jenkins and reusing its code is difficult.

Also, I have noticed that Zuul has more than one provider for its test cases. We talked before about those that come from Tempest, but there are also simple jobs that run a couple of Ansible commands and consider that their test. I have added test models for those two cases for now. Later on we may add more if necessary.

**Comment by bregman-arie** [comment]
> @bregman-arie The test models are very similar in both systems, but I have given Zuul its own model because I have had a bad experience reusing code from Jenkins before. Zuul always shows some condition that diverges from Jenkins and reusing its code is difficult.
> 
> Also, I have noticed that Zuul has more than one provider for its test cases. We talked before about those that come from Tempest, but there are also simple jobs that run a couple of Ansible commands and consider that their test. I have added test models for those two cases for now. Later on we may add more if necessary.

I understand the experience point, the systems are indeed very different but tests is actually a point where they should be pretty much the same as both Zuul and Jenkins run the same test/automation frameworks - tempest, tobiko, ansible, ...
Let's try to balance between diverging where needed and sharing where we have such an opportunity. If we don't have a technical reason to diverge right now, let's try to share.

**Comment by jsanemet** [comment]
@bregman-arie I have just updated the request so that Zuul tests inherit from Jenkins tests.

**Review by amolkahat — APPROVED** [review]

**Pull request merged**